src/librustc_trans/trans/common.rs

```diff
@@ -563,28 +568,49 @@ pub struct BlockS<'blk, 'tcx: 'blk> {
     // AST node-id associated with this block, if any. Used for
     // debugging purposes only.
     pub opt_node_id: Option<ast::NodeId>,
-
-    // The function context for the function to which this block is
-    // attached.
-    pub fcx: &'blk FunctionContext<'blk, 'tcx>,
 }
 
-pub type Block<'blk, 'tcx> = &'blk BlockS<'blk, 'tcx>;
-
-impl<'blk, 'tcx> BlockS<'blk, 'tcx> {
+impl BlockS {
     pub fn new(llbb: BasicBlockRef,
                is_lpad: bool,
-               opt_node_id: Option<ast::NodeId>,
-               fcx: &'blk FunctionContext<'blk, 'tcx>)
-               -> Block<'blk, 'tcx> {
-        fcx.block_arena.alloc(BlockS {
+               opt_node_id: Option<ast::NodeId>)
+               -> BlockS {
+        BlockS {
             llbb: llbb,
             terminated: Cell::new(false),
             unreachable: Cell::new(false),
             is_lpad: is_lpad,
             opt_node_id: opt_node_id,
-            fcx: fcx
-        })
+        }
+    }
+
+    pub fn alloc<'r, 'blk, 'tcx>(self, fcx: &FunctionContext<'blk, 'tcx>) -> &'blk mut BlockS {
+        fcx.block_arena.alloc(self)
+    }
+
+    pub fn with<'r, 'blk, 'tcx>(&'blk self,
+                                fcx: &'r mut FunctionContext<'blk, 'tcx>)
+                                -> Block<'r, 'blk, 'tcx> {
+        Block::new(self, fcx)
+    }
+}
+
+pub struct Block<'r, 'blk: 'r, 'tcx: 'blk> {
+    pub bl: &'blk BlockS,
+
+    // The function context for the function to which this block is
+    // attached.
+    pub fcx: &'r mut FunctionContext<'blk, 'tcx>,
```

Comment:
Since every block in a BlockS is in the same function, why do they each need their own reference to a function context? Seems like leaving it in BlockS is better?